File: Framework/CSharp/Framework/Framework/Web/SmartUrlParameter.cs
/*
 * 项目：Smartkernel.Framework
 * 作者：曹艳白
 * 邮箱：caoyanbai@139.com
 */
using Smartkernel.Framework.Text;
using System.Collections.Generic;
using System.Text;

namespace Smartkernel.Framework.Web
{
    /// <summary>
    /// UrlParameter
    /// </summary>
    public static class SmartUrlParameter
    {
        /// <summary>
        /// 编码参数对
        /// </summary>
        /// <param name="parameters">参数对列表</param>
        /// <param name="encoding">字符编码</param>
        /// <returns>编码结果</returns>
        public static string ToOne(Dictionary<string, string> parameters, Encoding encoding = null)
        {
            encoding = encoding ?? SmartEncoding.Default;
            var result = "";
            var enumerator = parameters.GetEnumerator();

            while (enumerator.MoveNext())
            {
                var kv = enumerator.Current;
                result += SmartHex.ToHex(kv.Key.ToLower(), encoding) + "=" + SmartHex.ToHex(kv.Value.ToLower(), encoding) + "|";
            }

            return SmartHex.ToHex(result.TrimEnd(new[] { '|' }), encoding);
        }

        /// <summary>
        /// 解码Url参数
        /// </summary>
        /// <param name="data">编码的参数</param>
        /// <param name="encoding">字符编码</param>
        /// <returns>结果之后的键值对</returns>
        public static Dictionary<string, string> ToAll(string data, Encoding encoding = null)
        {
            encoding = encoding ?? SmartEncoding.Default;
            var parameters = new Dictionary<string, string>();
            var output = SmartHex.FromHex(data, encoding);
            var kvs = output.Split(new[] { '|' });
            for (var i = 0; i < kvs.Length; i++)
            {
                var kv = kvs[i].Split(new[] { '=' });
                parameters.Add(SmartHex.FromHex(kv[0], encoding), SmartHex.FromHex(kv[1], encoding));
            }
            return parameters;
        }

        /// <summary>
        /// 查询一个参数（不存在则返回null）
        /// </summary>
        /// <param name="data">编码的数据</param>
        /// <param name="key">键</param>
        /// <param name="encoding">字符编码</param>
        /// <returns>值</returns>
        public static string Query(string data, string key, Encoding encoding = null)
        {
            encoding = encoding ?? SmartEncoding.Default;
            try
            {
                var parameters = ToAll(data, encoding);

                string value = null;

                foreach(var kv in parameters)
                {
                    if (kv.Key.ToLower().ToString() == key.ToLower())
                    {
                        value = kv.Value;
			break;
                    }
                }
                return value;
            }
            catch
            {
                return null;
            }
        }
    }
}

File: Framework/Php/framework/framework/web/smarturlparameter.php
<?php

/*
 * 项目：Smartkernel.Framework
 * 作者：曹艳白
 * 邮箱：caoyanbai@139.com
 */

namespace Smartkernel\Framework\Web {

    use Smartkernel\Framework\Text\SmartEncoding;
    use Smartkernel\Framework\Text\SmartHex;
    use Smartkernel\Framework\SmartString;
    use \Exception;

/*
     * UrlParameter
     */

    class SmartUrlParameter {
        /*
         * 编码参数对
         * 
         * @param array $parameters，参数对列表
         * @param string $encoding，字符编码
         * @return string，编码结果
         * 
         */

        public static function toOne($parameters, $encoding = null) {
            $encoding = $encoding == null ? SmartEncoding::DefaultValue : $encoding;
            $result = "";

            foreach ($parameters as $key => $value) {
                $result .= (SmartHex::toHex(SmartString::toLower($key), $encoding) . "=" . SmartHex::toHex(SmartString::toLower($value), $encoding) . "|");
            }
            return SmartHex::toHex(SmartString::trimEnd($result, "|"), $encoding);
        }

        /*
         * 解码Url参数
         * 
         * @param string $data，编码的参数
         * @param string $encoding，字符编码
         * @return array，结果之后的键值对
         * 
         */

        public static function toAll($data, $encoding = null) {
            $encoding = $encoding == null ? SmartEncoding::DefaultValue : $encoding;
            $parameters = array();
            $output = SmartHex::fromHex($data, $encoding);
            $kvs = SmartString::split($output, "|");
            $length = count($kvs);
            for ($i = 0; $i < $length; $i++) {
                $kv = SmartString::split($kvs[$i], "=");
                $parameters[SmartHex::fromHex($kv[0], $encoding)] = SmartHex::fromHex($kv[1], $encoding);
            }
            return $parameters;
        }

        /*
         * 查询一个参数（不存在则返回null）
         * 
         * @param string $data，编码的数据
         * @param string $key，键
         * @param string $encoding，字符编码
         * @return string，值
         * 
         */

        public static function query($data, $key, $encoding = null) {
            try {
                $parameters = self::toAll($data, $encoding);

                $value = null;

                foreach ($parameters as $k => $v) {
                    if (SmartString::toLower($k) == SmartString::toLower($key)) {
                        $value = $v;
                        break;
                    }
                }
                return $value;
            } catch (Exception $err) {
                return null;
            }
        }

    }

}
